app/lib/service/entrypoint/_messages.dart

```diff
@@ -0,0 +1,265 @@
+  ///
+  /// ```js
+  /// {
+  ///   ready: {
+  ///     requestSendPort: <port>,
+  ///   },
+  /// }
+  /// ```
+  ///
+  /// ### `debug` message, converted to [DebugMessage]
+  ///
+  /// ```js
+  /// {
+  ///   debug: {
+  ///     text: <string>,
+  ///   }
+  /// }
+  /// ```
+  ///
+  /// ### `request` message, converted to [RequestMessage]
+  ///
+  /// ```js
+  /// {
+  ///   request: {
+  ///     kind: <string>,
+  ///     payload: <object>,
+  ///     replyPort: <port>,
+  ///   }
+  /// }
+  /// ```
+  ///
+  /// ### `reply` message, converted to [ReplyMessage]
+  ///
+  /// ```js
+  /// {
+  ///   reply: {
+  ///     error: <string>,
+  ///     result: <object>,
+  ///   }
+  /// }
+  /// ```
+  static Message decodeFromMap(Map<String, dynamic> map) {
+    if (map.keys.length != 1) {
+      throw FormatException(
+          'Expected only a single key, got "${map.keys.join(', ')}"');
+    }
+    final key = map.keys.single;
+    final inner = map[key] as Map<String, dynamic>;
+    switch (key) {
+      case 'entry':
+        return EntryMessage(
+          protocolSendPort: inner['protocolSendPort'] as SendPort,
+          aliveSendPort: inner['aliveSendPort'] as SendPort,
+        );
+      case 'ready':
+        return ReadyMessage(
+          requestSendPort: inner['requestSendPort'] as SendPort?,
+        );
+      case 'request':
+        return RequestMessage(
+          inner['kind'] as String,
+          inner['payload'] as Object,
+          inner['replyPort'] as SendPort,
+        );
+      case 'reply':
+        return ReplyMessage._(
+          error: inner['error'] as String?,
+          result: inner['result'],
+        );
+      case 'debug':
+        return DebugMessage(inner['text'] as String);
+    }
+    throw FormatException('Unknown key: "$key".');
+  }
+
+  Map<String, dynamic> encodeAsJson();
+}
+
+/// Initializing message send from the controller isolate to the new isolate.
+class EntryMessage extends Message {
+  /// Port to which the isolate must be send its messages, starting with the
+  /// [ReadyMessage] when all the initialization is done after starting it.
+  ///
+  /// The isolate must use this port to send [RequestMessage] for cross-isolate
+  /// communication.
+  ///
+  /// The isolate may use this port to send [DebugMessage] when special logging
+  /// is needed.
+  final SendPort protocolSendPort;
+
+  /// Port to which any object may be sent periodically. After the first message
+  /// is sent, on each message, a timer will be set, and if no other message arrives,
+  /// the isolate will be killed and a new one will be started. The Timer's duration
+  /// is set at the isolate initialization and may depend on the task that the isolate
+  /// is doing.
+  ///
+  /// These message tell the isolate owner that the isolate isn't stuck.
+  /// In particular, this aims to allow the `IsolateRunner` to terminate
+  /// the isolate if it gets stuck in an infinite-loop.
+  final SendPort aliveSendPort;
```

Comment:
Why do we need two ports? We can cleanup later, I'm fine with that.

But I'm just asking: Could we do one, since we're already differentiating after deserializing messages sent on the `protocolSendPort` ?

Reply:
Yes we can (and shall) merge them eventually (or remove the aliveSendPort and the autokill timer)